newdle/api.py

```diff
@@ -200,6 +202,21 @@ def get_my_newdles():
     return MyNewdleSchema(many=True).jsonify(newdle)
 
 
+@api.route('/newdles/participating')
+def get_newdles_participating():
+    newdle = (
+        Participant.query.filter_by(auth_uid=g.user['uid'])
+        .join(Participant.newdle)
+        .order_by(
+            Participant.answers != '{}',
```

Comment:
There's one case we forgot: A newdle you didn't answer, but that has been finalized in the meantime. Adding a `Newdle.final_dt.isnot(None)` sort criterion at the very beginning should fix this.

Reply:
Thanks. I wasn't counting for expired newdles.
The criterion I'm now using is:
- show newdles without an answer and final date first (require attention)
- then newdles with a final date in the future (sorted by recent ones)
- then newdles without a final date (on-going)
- then past ones
> (unless we want to hide them by default, maybe behind a "show past" button - but that'd be a separate task). @ThiefMaster 

I ended up updating the sorting method to the following:
```
.order_by(
    or_(Participant.answers != '{}', Newdle.final_dt.isnot(None)),
    or_(Newdle.final_dt <= date.today(), Newdle.final_dt.is_(None)),
    Newdle.final_dt.isnot(None),
    Newdle.final_dt,
    Newdle.id.desc(),
)
```

Starting from the bottom, the idea is:
- Set up the base order to use the id descending, as previously
- Order the newdles by their final date ascending, in order to keep upcoming newdle dates sorted by their start
- Push newdles without a final date to the top (the on-going, to be determined ones)
- Push the upcoming, non-expired, newdles to the top (the relevant ones). However, ignore newdles without any final date. Because this comparison sets me the Nulls last.
- Show the newdles that require your immediate attention, without any final answers, first.

I was trying to avoid using the somewhat 'ugly' or_ or and_, but date comparisons were affecting the nulls. Nevertheless, I'm open to any suggestions for an alternative.
I will commit soon with a test that validates this approach.

Reply:
For `or_`/`and_ `vs `|`/`&` I usually decide like this:

- does it fit in one line? `|` and `&` it is (with parentheses around each of the operands)
- does it need to be wrapped? `and_` and `or_` are much more readable!

I think in this case using the operators wouldn't be so bad to use them..

Reply:
Technicalities aside, should we consider doing some grouping in the UI? Right now everything is in the same list, so it doesn't look super amazing if there's more than a few entries.

And if we sent the different categories in separate lists (ie an object as the top-level response item), we could simplify some of the sorting logic.

cc @indico/development-team

Reply:
I agree with the grouping. The sorting logic is becoming a bit too complex. Especially since the solution I provided sacrifices sorting past newdles correctly. While their priority is low, separate lists keep the view clean and easier to maintain.